File: apps/backend/src/cli/tools/phase.go
// src/cli/tools/phase.go
package tools

import (
	"io"

	"github.com/lucas/oraculo/apps/backend/src/db"
	"github.com/lucas/oraculo/apps/backend/src/domain"
	"github.com/lucas/oraculo/apps/backend/src/output"
	"github.com/spf13/cobra"
)

func newPhaseCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "phase",
		Short: "Manage session phases",
	}
	cmd.AddCommand(newPhaseCompleteCmd())
	return cmd
}

func newPhaseCompleteCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "complete <phase>",
		Short: "Complete a phase (reads JSON data from stdin)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			phase := args[0]
			w := cmd.OutOrStdout()
			database := dbFromContext(cmd.Context())
			sessionID, _ := cmd.Flags().GetString("session")

			data, err := io.ReadAll(cmd.InOrStdin())
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			dataStr := string(data)
			if dataStr == "" {
				dataStr = "{}"
			}

			sessionStore := db.NewSessionStore(database)
			if err := sessionStore.CompletePhase(sessionID, phase, dataStr); err != nil {
				output.WriteError(w, err)
				return err
			}

			// Determine next phase
			session, _ := sessionStore.Get(sessionID)
			next, _ := sessionStore.CurrentPhase(sessionID)
			sessionClosed := session.Status == domain.SessionCompleted

			result := map[string]any{
				"phase":     phase,
				"completed": true,
				"next":      next,
			}
			if sessionClosed {
				result["session_closed"] = true
			}
			return output.WriteJSON(w, result)
		},
	}
	cmd.Flags().String("session", "", "Session ID")
	cmd.MarkFlagRequired("session")
	return cmd
}

File: apps/orchestrator/src/cli/tools/phase.ts
import { Command } from "commander";
import { SessionStore } from "../../db/stores/session-store.ts";
import { PHASES } from "../../domain/enums.ts";
import type { SessionType } from "../../domain/enums.ts";
import { writeJSON, writeError } from "../../utils/output.ts";
import { dbFromCmd } from "../context.ts";

export function phaseCommands(): Command {
  const cmd = new Command("phase").description("Manage session phases");

  cmd.addCommand(phaseCompleteCmd());

  return cmd;
}

function phaseCompleteCmd(): Command {
  return new Command("complete")
    .description("Complete a phase (reads JSON data from stdin)")
    .argument("<phase>", "Phase name")
    .requiredOption("--session <id>", "Session ID")
    .action(async (phase: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const sessionStore = new SessionStore(db);

        let data = await Bun.stdin.text();
        if (!data.trim()) {
          data = "{}";
        }

        sessionStore.CompletePhase(opts.session, phase, data);

        // Determine next phase
        const session = sessionStore.Get(opts.session);
        const allPhases: string[] =
          PHASES[session.type as SessionType] ?? [];
        const completedPhases = db
          .query<{ name: string }, [string]>(
            "SELECT phase as name FROM session_phases WHERE session_id = ?",
          )
          .all(session.id)
          .map((p) => p.name);

        let next = "";
        for (const ph of allPhases) {
          if (!completedPhases.includes(ph)) {
            next = ph;
            break;
          }
        }

        const result: Record<string, unknown> = {
          phase,
          completed: true,
          next,
        };

        if (session.status === "completed") {
          result.session_closed = true;
        }

        writeJSON(result);
      } catch (err) {
        writeError(err);
      }
    });
}
